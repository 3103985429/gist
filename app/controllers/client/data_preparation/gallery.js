const winston = require('winston');
const { get } = require('lodash');
const { VIEW_TYPE_GALLERY, GALLERY_STYLE_IMAGE } = require('../../../config/views.config');

const { Lazy_Shared_ProcessedRowObject_MongooseContext } = require('../../../models/processed_row_objects');
const { calculateGalleryPageRanges } = require('../config');
const func = require('../func');
const { processDocs } = require('./data_prep_helpers/gallery.helpers');
const { getGallerySort } = require('./data_prep_helpers/sort.helpers');
const { getGalleryProject } = require('./data_prep_helpers/project.helpers');
const {
    addDefaultAggregationOperators,
    aggregateProcessedRowObjects,
} = require('./data_prep_helpers/aggregation.helpers');

module.exports = { bindData };

async function bindData(dataSource, options) {
    const { filterObj, limit, onPageNum, routePath_base, skipNResults } = options;
    const galleryViewSettings = dataSource.fe_views.views.gallery;
    const { galleryStyle } = galleryViewSettings;
    const { Model: processedRowObjects_mongooseModel } = Lazy_Shared_ProcessedRowObject_MongooseContext(dataSource._id);

    const wholeFilteredSetAggregationOperators = addDefaultAggregationOperators(dataSource, options);

    const groupQuery = {
        $group: {
            _id: null,
            count: { $sum: 1 },
        },
    };

    winston.debug('Getting nonPagedCount');

    const countWholeFilteredSetAggregationOperators = wholeFilteredSetAggregationOperators.concat([groupQuery]);
    const nonPagedResults = await aggregateProcessedRowObjects(processedRowObjects_mongooseModel, countWholeFilteredSetAggregationOperators);
    const nonPagedCount = get(nonPagedResults, [0, 'count'], 0);

    winston.debug('Found %i non-paged docs', nonPagedCount);

    const sort = getGallerySort(dataSource, options);
    const additionalProject = getGalleryProject(dataSource, options);
    const project = {
        _id: 1,
        pKey: 1,
        srcDocPKey: 1,
        ...additionalProject,
    };

    const pagedDocs_aggregationOperators = wholeFilteredSetAggregationOperators.concat([
        { $project: project },
        { $sort: sort },
        { $skip: skipNResults },
        { $limit: limit },
    ]);

    winston.debug('Getting paged docs');

    const docs = await aggregateProcessedRowObjects(processedRowObjects_mongooseModel, pagedDocs_aggregationOperators);
    const processedResults = processDocs(docs, dataSource, VIEW_TYPE_GALLERY, galleryViewSettings, options);

    const pageRanges = calculateGalleryPageRanges(nonPagedCount);
    const { fe_image } = dataSource;
    const hasImages = galleryStyle === GALLERY_STYLE_IMAGE && fe_image.gallery && fe_image.field.length;

    winston.debug('Returning %i docs', processedResults.length);

    const data = {
        data: processedResults,
        included: {
            dataSource: dataSource,
        },
        meta: {
            numberOfResults: processedResults.length,
            numPages: Math.ceil(nonPagedCount / limit),
            nonpagedCount: nonPagedCount,
            pageRanges,
            onPageNum,
            routePath_base,
            filterObj,
            hasImages,
        },
    };

    winston.debug('Done preparing data for gallery');

    return data;
}
