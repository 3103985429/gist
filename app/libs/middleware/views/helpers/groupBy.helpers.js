const { cond, constant, eq, has, stubTrue } = require('lodash');

const { RealColumnNameFromHumanReadableColumnName } = require('../../../datasources/imported_data_preparation');
const {
    SETTING_DEFAULT_GROUP_BY_DURATION,
    SETTING_DEFAULT_GROUP_BY_RANGE,
    SETTING_DEFAULT_GROUP_BY_COLUMN_NAME,
    SETTING_DEPENDS_ON_SORT_BY,
} = require('../../../../config/views.config');
const { fieldOverrideIfExists } = require('../../../../controllers/client/func');
const { equals } = require('../../../../utils/helpers');

const TYPE_DATE = 'date';
const TYPE_NUMBER = 'number';
const TYPE_STRING = 'string';

module.exports = {
    getDefaultGroupByColumnName,
    getGroupByTypeFromSortBy,
    invalidateGroupBy,
};

/**
 * @param {String} groupByType
 * @param {Object} dataSourceDescription
 * @param {String?} defaultGroupByDuration
 * @param {String?} defaultGroupByRange
 * @param {String?} defaultGroupByColumnName
 * @returns {String|undefined}
 */
function getDefaultGroupByColumnName(
    groupByType,
    dataSourceDescription,
    { defaultGroupByDuration, defaultGroupByRange, defaultGroupByColumnName }
) {
    if (groupByType === TYPE_DATE) {
        return defaultGroupByDuration;
    }

    if (groupByType === TYPE_NUMBER) {
        return defaultGroupByRange;
    }

    return fieldOverrideIfExists(defaultGroupByColumnName, dataSourceDescription);
}

/**
 * @param {Boolean} isDate
 * @param {Boolean} isNumber
 * @param {Object[]} masterViewSettings
 * @returns {String|null}
 */
function getGroupByTypeFromSortBy({ isDate, isNumber } = {}, masterViewSettings) {
    const findSetting = settingName => masterViewSettings.find(({ name }) => name === settingName);
    const getType = cond([
        [() => isDate && findSetting(SETTING_DEFAULT_GROUP_BY_DURATION), constant(TYPE_DATE)],
        [() => isNumber && findSetting(SETTING_DEFAULT_GROUP_BY_RANGE), constant(TYPE_NUMBER)],
        [
            () => has(findSetting(SETTING_DEFAULT_GROUP_BY_COLUMN_NAME), SETTING_DEPENDS_ON_SORT_BY),
            constant(TYPE_STRING),
        ],
        [stubTrue, constant(null)],
    ]);

    return getType();
}

function invalidateGroupBy(
    groupBy,
    groupByType,
    dataSourceDescription,
    { fieldsNotAvailableAsGroupByColumns = [], durationsAvailableForGroupBy = [], rangesAvailableForGroupBy = [] }
) {
    const { fe_excludeFields } = dataSourceDescription;
    const groupByRealColumnName = RealColumnNameFromHumanReadableColumnName(groupBy, dataSourceDescription, true);
    const stringGroupBy = () =>
        eq(fe_excludeFields[groupByRealColumnName], false) && !fieldsNotAvailableAsGroupByColumns[groupByRealColumnName]
            ? groupBy
            : undefined;
    const dateGroupBy = () => (durationsAvailableForGroupBy.includes(groupBy) ? groupBy : undefined);
    const numberGroupBy = () => (rangesAvailableForGroupBy.includes(groupBy) ? groupBy : undefined);
    const getDefaultGroupBy = cond([
        [equals(TYPE_STRING), stringGroupBy],
        [equals(TYPE_DATE), dateGroupBy],
        [equals(TYPE_NUMBER), numberGroupBy],
        [stubTrue, constant(groupBy)],
    ]);

    return getDefaultGroupBy(groupByType);
}
