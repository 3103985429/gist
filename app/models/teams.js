const integerValidator = require('mongoose-integer');
const winston = require('winston');
const { isEmpty } = require('lodash');

const mongoose_client = require('./mongoose_client');

const mongoose = mongoose_client.mongoose;
const Schema = mongoose.Schema;

const team_scheme = Schema({
    title: String,
    subdomain: { type: String, unique: true },
    description: String,
    font: { type: String, default: 'centrano2' },
    logo: String,
    logo_header: String,
    externalWebsite: String,
    colorPalette: {
        type: Array,
        default: [
            '#1B6DFC',
            '#0CB7FA',
            '#FA79FC',
            '#FB3705',
            '#F9D307',
            '#FA9007',
            '#7D5807',
            '#B2E606',
            '#069908',
            '#7B07FD',
        ],
    },
    superTeam: Boolean,
    admin: [{ type: Schema.Types.ObjectId, ref: 'User' }],
    datasourceDescriptions: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    pages: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    sites: [{ type: Schema.Types.ObjectId, ref: 'Website' }],

    isEnterprise: Boolean, //if true, means it has custom view - TODO consolidate with hasEnterpriseLicense
    customViewModal: Boolean, //if true, custom view has custom view settings modal - (rhodiumgroup)
    hasEnterpriseLicense: { type: Boolean, default: false },

    google_analytics: String,
    header_script: String,
    footer_script: String,
    brandColor: {
        label: { type: String, default: '#2A2A2A' },
        select: { type: String, default: '#2A2A2A' },
    },

}, { timestamps: true });

const modelName = 'Team';

team_scheme.plugin(integerValidator);

const deepPopulate = require('mongoose-deep-populate')(mongoose);

team_scheme.plugin(deepPopulate, { whitelist: ['datasourceDescriptions.author', 'datasourceDescriptions.updatedBy', 'datasourceDescriptions.createdAt'] });

team_scheme.pre('save', function (next) {
    this._wasNew = this.isNew;
    next();
});

team_scheme.methods.notifyNewTeamCreation = function () {
    if (this._wasNew) {
        this.populate('admin', function (err, docPopulatedWithAdmin) {
            if (err || isEmpty(docPopulatedWithAdmin.admin)) {
                winston.error('Team created with error err:' + err);
            }
        });
    }
};

team_scheme.statics.getTeamEntries = async (subdomain, user, vizName) => {
    const baseQuery = { subdomain };

    if (process.env.NODE_ENV === 'enterprise' && !process.env.SUBTEAMS) {
        baseQuery.subdomain = process.env.SUBDOMAIN;
    }

    const populateQuery = [
        {
            path: 'datasourceDescriptions',
            match: require('./descriptions').buildTeamPageQuery(user, vizName),
            select: 'description uid urls title importRevision updatedAt createdAt previous_datasets updatedBy author brandColor fe_views fe_filters.default banner connection',
            populate: [
                {
                    path: 'updatedBy',
                    select: 'firstName lastName',
                },
                {
                    path: 'author',
                    select: 'firstName lastName',
                },
                {
                    path: 'previous_datasets',
                    select: 'createdAt -_id',
                },
            ],
            options: {
                sort: {
                    createdAt: -1,
                },
            },
        },
        {
            path: 'pages',
            match: require('./pages').buildTeamPageQuery(user),
            select: 'pageTitle slug parsedSummary url thumbnail image updatedBy createdBy createdAt updatedAt createdAt',
            populate: [
                {
                    path: 'createdBy',
                    select: 'firstName lastName -_id',
                },
                {
                    path: 'website',
                    select: 'slug -_id',
                },
            ],
            options: {
                sort: {
                    createdAt: -1,
                },
            },
        },
    ];

    return team.find(baseQuery).populate(populateQuery);
};

const team = mongoose.model(modelName, team_scheme);

team.GetTeams = function (fn) {
    mongoose_client.WhenMongoDBConnected(function () {
        team.find({}, function (err, teams) {
            if (err) {
                fn(err);
            } else {
                fn(null, teams);
            }
        });
    });
};

module.exports = team;
