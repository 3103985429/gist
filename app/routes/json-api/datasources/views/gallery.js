const winston = require('winston');
const express = require('express');

const { addDataPrepOptions } = require('../../../../libs/middleware/views/add-data-prep-options');
const { bindData } = require('../../../../controllers/client/data_preparation/gallery');
const { handleError } = require('../../../../utils/requests');

const router = express.Router();

router.get('/graph-data', addDataPrepOptions, async (req, res) => {
    try {
        const results = await bindData(req.dataSource, req.dataPrepOptions);
        results.meta.imageMeta = req.imageMeta;

        winston.debug('Returning results');
        res.json(results);
    } catch (err) {
        winston.error('Error getting bind data for json-api datasource gallery data', err);
        handleError(err, res);
    }
});

module.exports = router;
