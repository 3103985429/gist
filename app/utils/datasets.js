const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
