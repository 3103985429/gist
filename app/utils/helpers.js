const { isEqual, trim, toString } = require('lodash');

module.exports.equals = a => b => isEqual(a, b);

module.exports.isEmbed = (query) => query.embed === 'true' || query.embed === true;

module.exports.formatColumnName = (value) => trim(toString(value).replace(/\./g, '_'));

module.exports.formatRawValues = (value) => trim(toString(value).replace(/’/g, '\'').replace(/”/g, '"'));

module.exports.addSlugIndex = (slugs, item, tilteProp = 'title') => {
    if (!slugs.length) {
        return;
    }

    let i = 2;
    const slugList = slugs.map((item) => item.slug);
    const max = i + slugList.length + 1;

    for (i; i < max; i++) {
        const newSlug = `${item.slug}-${i}`;

        if (!slugList.includes(newSlug)) {
            item.slug = newSlug;
            item[tilteProp] = `${item[tilteProp]} ${i}`;

            break;
        }
    }
};
