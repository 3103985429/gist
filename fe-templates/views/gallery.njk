{# TODO: why set all these variables like this? easier initial transition from the existing gallery templates.
    lots of opportunity to precalculate this stuff in the data_prep controller and slim down the templates. #}

{% set dataSource = included.dataSource %}
{% set coercionScheme = dataSource.raw_rowObjects_coercionScheme %}
{% set viewOptions = dataSource.fe_viewOptions %}
{% set array_source_key = dataSource.uid %}
{% set fieldKey_objectTitle = options.fieldKey_objectTitle %}
{% set galleryStyle = options.galleryStyle %}
{% set secondaryColumnTileDisplay = options.secondaryColumnTileDisplay %}
{% set secondaryColumnTileShow = options.secondaryColumnTileShow %}
{% set nonpagedCount = meta.nonpagedCount %}
{% set scrapedImages = meta.imageMeta.scrapedImages %}
{% set sharedPage = options.sharedPage %}
{% set routePath_base = options.routePath_base %}
{% set embedded = options.embedded %}
{% set isPreview = options.preview %}
{% set revision = options.revision %}
{% set hasThumbs = meta.imageMeta.hasThumbs %}
{% set galleryViewSettings = dataSource.fe_views.views.gallery %}
{% set displaySortFieldText = galleryViewSettings.displaySortFieldText %}
{% set displayFieldTitles = galleryViewSettings.displayFieldTitles %}
{% set team = dataSource._team if dataSource._team else null %}
{% set subdomain = team.subdomain | lower %}
{% set datasetId = dataSource._id %}
{% set viewType = 'gallery' %}
{% set sortBy = options.sortBy %}
{% set sortBy_realColumnName = options.sortBy_realColumnName %}
{% set defaultSortByColumnName_humanReadable = options.defaultSortByColumnName_humanReadable %}
{% set secondaryCol = options.secondaryCol %}
{% set filterObj = options.filterObj %}
{% set imageCaption = true %}

<ul class="gallery-grid gallery-single">
    {% for doc in data %}
        {% if sortBy_realColumnName %}
            {% set textField = sortBy_realColumnName %}
        {% else %}
            {% set textField = defaultSortByColumnName_humanReadable %}
        {% endif %}

        {% if textField === fieldKey_objectTitle %}
            {% set displaySortFieldText = false %}
        {% endif %}

        {% if galleryStyle === 'image' and hasThumbs and doc.nestedImagePath %}
            {% include 'views/partials/gallery_item_image.njk' %}
        {% else %}
            {% include 'views/partials/gallery_item_text.njk' %}
        {% endif %}
    {% endfor %}
</ul>
