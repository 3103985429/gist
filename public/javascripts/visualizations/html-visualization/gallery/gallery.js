/* global arrays, jQuery, nunjucks */

/**
 * External namespace for arrays classes
 * @external arrays
 */

/**
 * Creates a new Gallery
 * @param {Object} options
 * @constructor
 * @extends HTMLVisualization
 */
arrays.Gallery = function () {

    // Call the parent constructor, making sure (using Function#call)
    // that "this" is set correctly during the call
    arrays.HTMLVisualization.call(this);

    this._templateName = 'views/gallery.njk';
    this._meta = {};
    this._included = {};
};

// Create a Gallery.prototype object that inherits from HTMLVisualization.prototype.
arrays.Gallery.prototype = Object.create(arrays.HTMLVisualization.prototype);

// Set the "constructor" property to refer to Gallery
arrays.Gallery.prototype.constructor = arrays.Gallery;

/**
 * Initialize the Gallery
 * @param {Object} data - All data required to render this visualization
 * @param {Object} options
 * @public
 * @extends HTMLVisualization
 * @returns {Gallery}
 */
arrays.Gallery.prototype.init = function (data, options) {
    arrays.HTMLVisualization.prototype.init.call(this, data, options);

    this.setMeta(data.meta);
    this.setIncluded(data.included);

    return this;
};

/**
 * Renders a Gallery
 * @private
 */
arrays.Gallery.prototype._render = function () {
    const isMobile = arrays.isMobileDevice();
    const html = nunjucks.render(this._templateName, {
        data: this._data,
        meta: this._meta,
        included: this._included,
        options: this._options,
    });

    const titleMaxLines = 3;
    const subtitleMaxLines = isMobile ? 2 : 1;

    this._wrapper.append(html);

    const galleryTitleElement = $('.gallery-title');
    const gallerySecondaryTitleElement = $('.gallery-secondary-title');
    const titleLineHeight = parseInt(galleryTitleElement.css('line-height').replace('px', ''), 10);

    if (gallerySecondaryTitleElement.length !== 0) {
        const secondaryTitleLineHeight = parseInt(gallerySecondaryTitleElement.css('line-height').replace('px', ''), 10);
        gallerySecondaryTitleElement.shave(subtitleMaxLines * secondaryTitleLineHeight);
    }

    galleryTitleElement.shave(titleMaxLines * titleLineHeight);
};

/**
 * Set meta
 * @param {Object} meta
 * @public
 */
arrays.Gallery.prototype.setMeta = function (meta) {
    this._meta = arrays.clone(meta);
};

/**
 * Set included
 * @param {Object} included
 * @public
 */
arrays.Gallery.prototype.setIncluded = function (included) {
    this._included = arrays.clone(included);
};
