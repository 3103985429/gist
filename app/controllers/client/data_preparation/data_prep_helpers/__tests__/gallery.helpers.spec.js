import { toArray } from 'lodash';

import { processDocs } from '../gallery.helpers';
import { pokemon1 } from '../../../../../../internals/testing/backend/fixtures/datasets';
import { team1 } from '../../../../../../internals/testing/backend/fixtures/teams';
import * as processedRowObjects from '../../../../../../internals/testing/backend/fixtures/pokemon1ProcessedRowObjects';


describe('gallery helpers', () => {
    describe('processDocs', () => {
        let options, dataSet;

        beforeEach(() => {
            options = {
                sortBy: 'Name',
                sortBy_realColumnName: 'Name',
            };
            dataSet = Object.assign({}, pokemon1, { _team: team1 });
        });

        it('should return docs', () => {
            const usePercent = false;
            const result = processDocs(toArray(processedRowObjects), dataSet, 'gallery', dataSet.fe_views.views.gallery, options, usePercent);
            expect(result).toMatchSnapshot();
        });
    });
});
