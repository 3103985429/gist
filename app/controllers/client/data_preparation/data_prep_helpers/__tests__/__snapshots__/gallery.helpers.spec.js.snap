// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gallery helpers processDocs should return docs 1`] = `
Array [
  Object {
    "_id": "5b05346ad7a86407e5ef783d",
    "pKey": "2",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/2/21/001Bulbasaur.png/250px-001Bulbasaur.png",
      "Max CP": 1079,
      "Max HP": 83,
      "Name": "Bulbasaur",
      "Pokemon No_": "1",
      "Type 1": "Grass",
      "Type 2": "Poison",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
  Object {
    "_id": "5b05346ad7a86407e5ef783e",
    "pKey": "4",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/7/73/002Ivysaur.png/250px-002Ivysaur.png",
      "Max CP": 1643,
      "Max HP": 107,
      "Name": "Ivysaur",
      "Pokemon No_": "2",
      "Type 1": "Grass",
      "Type 2": "Poison",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
  Object {
    "_id": "5b05346ad7a86407e5ef783f",
    "pKey": "6",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/a/ae/003Venusaur.png/250px-003Venusaur.png",
      "Max CP": 2598,
      "Max HP": 138,
      "Name": "Venusaur",
      "Pokemon No_": "3",
      "Type 1": "Grass",
      "Type 2": "Poison",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7840",
    "pKey": "8",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/7/73/004Charmander.png/250px-004Charmander.png",
      "Max CP": 962,
      "Max HP": 73,
      "Name": "Charmander",
      "Pokemon No_": "4",
      "Type 1": "Fire",
      "Type 2": "None",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7841",
    "pKey": "10",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/4/4a/005Charmeleon.png/250px-005Charmeleon.png",
      "Max CP": 1568,
      "Max HP": 103,
      "Name": "Charmeleon",
      "Pokemon No_": "5",
      "Type 1": "Fire",
      "Type 2": "None",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7842",
    "createdAt": "2018-05-23T09:29:41.659Z",
    "pKey": "12",
    "published": false,
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/7/7e/006Charizard.png/250px-006Charizard.png",
      "Max CP": 2620,
      "Max HP": 135,
      "Name": "Charizard",
      "Pokemon No_": "6",
      "Type 1": "Fire",
      "Type 2": "Flying",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
    "updatedAt": "2018-05-23T09:29:51.572Z",
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7843",
    "createdAt": "2018-05-23T09:29:42.086Z",
    "pKey": "14",
    "published": false,
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/3/39/007Squirtle.png/250px-007Squirtle.png",
      "Max CP": 1015,
      "Max HP": 81,
      "Name": "Squirtle",
      "Pokemon No_": "7",
      "Type 1": "Water",
      "Type 2": "None",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
    "updatedAt": "2018-05-23T09:29:52.210Z",
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7844",
    "createdAt": "2018-05-23T09:29:42.541Z",
    "pKey": "16",
    "published": false,
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/0/0c/008Wartortle.png/250px-008Wartortle.png",
      "Max CP": 1594,
      "Max HP": 105,
      "Name": "Wartortle",
      "Pokemon No_": "8",
      "Type 1": "Water",
      "Type 2": "None",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
    "updatedAt": "2018-05-23T09:29:54.240Z",
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7845",
    "pKey": "18",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/0/02/009Blastoise.png/250px-009Blastoise.png",
      "Max CP": 2560,
      "Max HP": 137,
      "Name": "Blastoise",
      "Pokemon No_": "9",
      "Type 1": "Water",
      "Type 2": "None",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
]
`;
